### PR TITLE
Add preliminary check before registering an analyzer

### PR DESCRIPTION
**Pull request by zsolt-kolbay-sonarsource**

### CODE REVIEW

File: analyzers/src/SonarAnalyzer.Common/DiagnosticAnalyzer/SonarDiagnosticAnalyzer.cs [hunk 1]
```diff
@@ -36,7 +37,21 @@ public sealed override void Initialize(RoslynAnalysisContext context)
             {
                 context.EnableConcurrentExecution();
             }
-            Initialize(new SonarAnalysisContext(context, SupportedDiagnostics));
+
+            if (ShouldRegisterAnalyzerPredicate != null)
+            {
+                context.RegisterCompilationStartAction(compilation =>
```
Comment (pavel-mikula-sonarsource) [inline analyzers/src/SonarAnalyzer.Common/DiagnosticAnalyzer/SonarDiagnosticAnalyzer.cs:43]: The lambda argument is not `compilation` but `startContext`

File: analyzers/src/SonarAnalyzer.Common/DiagnosticAnalyzer/SonarDiagnosticAnalyzer.cs [hunk 2]
```diff
@@ -27,6 +27,7 @@ public abstract class SonarDiagnosticAnalyzer : DiagnosticAnalyzer
         public static readonly string EnableConcurrentExecutionVariable = "SONAR_DOTNET_ENABLE_CONCURRENT_EXECUTION";
 
         protected virtual bool EnableConcurrentExecution => IsConcurrentExecutionEnabled();
+        protected virtual Func<Compilation, bool> ShouldRegisterAnalyzerPredicate => null;
```
Comment (pavel-mikula-sonarsource) [inline analyzers/src/SonarAnalyzer.Common/DiagnosticAnalyzer/SonarDiagnosticAnalyzer.cs:30]: The name is too long :)

File: analyzers/src/SonarAnalyzer.Common/DiagnosticAnalyzer/SonarDiagnosticAnalyzer.cs (hunk 1 above)
Comment (pavel-mikula-sonarsource) [inline analyzers/src/SonarAnalyzer.Common/DiagnosticAnalyzer/SonarDiagnosticAnalyzer.cs:43]: Will this work for actions that register their own CompilationStartAction inside this CompilationStartAction? Just out of curiosity

File: analyzers/src/SonarAnalyzer.Common/DiagnosticAnalyzer/SonarDiagnosticAnalyzer.cs (hunk 2 above)
Comment (pavel-mikula-sonarsource) [inline analyzers/src/SonarAnalyzer.Common/DiagnosticAnalyzer/SonarDiagnosticAnalyzer.cs:30]: Lambda is painful to use, and the double handling is kindof hacky.

What about new abstract class with abstract member to override, to make the contract more clear?

**Comment by martin-strecker-sonarsource** [comment]
@zsolt-kolbay-sonarsource This can be closed as the current implementation of the rule is just fine. I don't see why we need this additional indirection.

**Comment by zsolt-kolbay-sonarsource** [comment]
@martin-strecker-sonarsource 
Sorry for the confusion. I'll add a description to the PR. This is an experiment for adding registration checks to analyzers. E.g. if a rule checks WCF services then it should only be registered if the project references WCF libraries. This is to trim the analysis time by not running unnecessary analyzers on a project.

**Comment by martin-strecker-sonarsource** [comment]
What I'm saying is that the current master already doing the conditional check. There is no need to introduce `ShouldRegisterAnalyzerPredicate` as there is an idiomatic way how to do such things:
https://github.com/SonarSource/sonar-dotnet/blob/b36349a9bbef58e114f47446a57878fe10c74969/analyzers/src/SonarAnalyzer.Common/Rules/DateAndTimeShouldNotBeUsedasTypeForPrimaryKeyBase.cs#L45-L49

**Comment by zsolt-kolbay-sonarsource** [comment]
We're going to use `RegisterCompilationStartAction` for these kinds of checks, so I'm closing this PR.

**Pull request closed without merging**